### PR TITLE
Add initial support for distroless containers

### LINKED ISSUE #864: Enable inventory of Distroless images
**Describe the Feature**
Distroless provides runtime container images. It doesn't have a shell and therefore not as straightforward to enable inventorying via a binary. However, there is a manifest within the rootfs that can read, so inventorying can still be done.

**Use Cases**
See https://github.com/ossf/wg-digital-identity-attestation/issues/20
We also have difficulty running file level static analysis to get an SBoM for this kind of rootfs, so a deterministic method would be very useful.

**Implementation Changes**
See comment: https://github.com/ossf/wg-digital-identity-attestation/issues/20#issuecomment-724091084

The "host" method invokes a host command or script on the working directory without the `unshare -pf` which you would get from the `invoke_in_chroot`.

### PR DESCRIPTION
**Pull request by JamieMagee**
This PR adds support for extracting package information from [distroless containers](https://github.com/GoogleContainerTools/distroless).

As-is, I'm able to extract package names and versions from [all current tags](https://gcr.io/v2/distroless/tags/list). I'm also able to extract copyright information, but not had success mapping it to licenses. Please see the logs for:
- [gcr.io/distroless/base:latest](https://gist.github.com/JamieMagee/d9cf1ca579b230db504186cbae816f2d#file-distroless-base-latest-log)
- [gcr.io/distroless/dotnet:latest](https://gist.github.com/JamieMagee/d9cf1ca579b230db504186cbae816f2d#file-distroless-dotnet-latest-log)
- [gcr.io/distroless/python3:debug](https://gist.github.com/JamieMagee/d9cf1ca579b230db504186cbae816f2d#file-distroless-python3-debug-log)

Closes #864

### CODE REVIEW

**Comment by nishakm** [comment]
> as provided by Google specifically, not distroless containers in general)

What is the difference between distroless containers provided by the [distroless project](https://github.com/GoogleContainerTools/distroless) and distroless containers "in general"?

**Comment by JamieMagee** [comment]
When I say "distroless containers in general" I mean the concept of distroless containers. I've heard the term distroless used to describe minimal, barebones, containers with a single executable.

File: tern/analyze/default/container/single_layer.py [hunk 1]
```diff
@@ -144,6 +144,10 @@ def analyze_first_layer(image_obj, master_list, redo):
             # unmount
             rootfs.undo_mount()
             rootfs.unmount_rootfs()
+        # distroless images do not have a shell or binaries in the base layer
+        # we see what the OS is reported as instead
```
Comment (nishakm) [inline tern/analyze/default/container/single_layer.py:148]: If we can find `/var/lib/dpkg/status.d` on the filesystem then we can use that as an indication of a distroless base image.

File: tern/analyze/default/container/multi_layer.py [hunk 2]
```diff
@@ -67,6 +67,10 @@ def fresh_analysis(image_obj, curr_layer, shell, driver, envs):
                     envs)
         rootfs.undo_mount()
         rootfs.unmount_rootfs()
+    # distroless images do not have a command list
```
Comment (nishakm) [inline tern/analyze/default/container/multi_layer.py:70]: I don't think there's any way of collecting info from the things that get added on top of the first layer. My investigation has found that `/var/lib/dpkg/status.d` doesn't get updated in the second layer for all the distroless images besides the static one. Let's leave this change out for now until we can figure out an alternative.

**Review by nishakm — CHANGES_REQUESTED** [review]
I've asked for an alternative way to include a distroless "binary". The script only detects whether a file/directory in `base.yml` exists on the filesystem. If it does then one can go ahead and start invoking scripts. As for changes in `snippets.yml`, I've asked to belay that change for now.

File: tern/analyze/default/container/multi_layer.py (hunk 2 above)
Comment (JamieMagee) [reply tern/analyze/default/container/multi_layer.py:70]: This is not the experience that I found. When scanning `gcr.io/distroless/dotnet:latest` I find 4 layers, each with packages:

<details>
<summary>Details</summary>

```
Docker image: gcr.io/distroless/dotnet@sha256:8a4099dee4df6f856bad39bbcf598255fa30061cd8227bbd08edda147999f54b:
        Layer 1:
                warning: No known shell found in image. Cannot invoke commands

                info: Found 'Distroless' in /etc/os-release.
                warning: No listing method for 'files'. Additional analysis may be required.
No listing method for 'licenses'. Additional analysis may be required.
No listing method for 'srcs'. Additional analysis may be required.

        File licenses found in Layer:  None
        Packages found in Layer:  base-files-9.9+deb9u13, netbase-5.4, tzdata-2021a-0+deb9u1
        Licenses found in Layer:  None
------------------------------------------------

        Layer 2:
                warning: No created_by information for layer.
                warning: No listing method for 'files'. Additional analysis may be required.
No listing method for 'licenses'. Additional analysis may be required.
No listing method for 'srcs'. Additional analysis may be required.

        File licenses found in Layer:  None
        Packages found in Layer:  libc6-2.24-11+deb9u4, libssl1.1-1.1.0l-1~deb9u3, openssl-1.1.0l-1~deb9u3
        Licenses found in Layer:  None
------------------------------------------------

        Layer 3:
                warning: No created_by information for layer.
                warning: No listing method for 'files'. Additional analysis may be required.
No listing method for 'licenses'. Additional analysis may be required.
No listing method for 'srcs'. Additional analysis may be required.

        File licenses found in Layer:  None
        Packages found in Layer:  libgcc1-1:6.3.0-18+deb9u1, libgomp1-6.3.0-18+deb9u1, libstdc++6-6.3.0-18+deb9u1
        Licenses found in Layer:  None
------------------------------------------------

        Layer 4:
                warning: No created_by information for layer.
                error: /bin/cat: /usr/share/doc/libcomerr2/copyright: No such file or directory
/bin/cat: /usr/share/doc/libcurl3/copyright: No such file or directory
/bin/cat: /usr/share/doc/libicu57/copyright: No such file or directory
/bin/cat: /usr/share/doc/libidn11/copyright: No such file or directory
/bin/cat: /usr/share/doc/libssl1.0.2/copyright: No such file or directory
/bin/cat: /usr/share/doc/libunistring0/copyright: No such file or directory

                warning: No listing method for 'files'. Additional analysis may be required.
No listing method for 'licenses'. Additional analysis may be required.
No listing method for 'srcs'. Additional analysis may be required.

        File licenses found in Layer:  None
        Packages found in Layer:  curl-7.52.1-5+deb9u13, libacl1-2.2.52-3+b1, libattr1-1:2.4.47-2+b2, libbz2-1.0-1.0.6-8.1, libcomerr2-1.44.5-1~bpo9+1, libcurl3-7.52.1-5+deb9u13, libdb5.3-5.3.28-12+deb9u1, libffi6-3.2.1-6, libgcrypt20-1.7.6-2+deb9u3, libgmp10-2:6.1.2+dfsg-1, libgnutls30-3.5.8-5+deb9u5, libgpg-error0-1.35-1~bpo9+1, libgssapi-krb5-2-1.15-1+deb9u2, libhogweed4-3.3-1+b2, libicu57-57.1-6+deb9u4, libidn11-1.33-1+deb9u1, libidn2-0-2.0.5-1~bpo9+1, libk5crypto3-1.15-1+deb9u2, libkeyutils1-1.5.9-9, libkrb5-3-1.15-1+deb9u2, libkrb5support0-1.15-1+deb9u2, libldap-2.4-2-2.4.47+dfsg-3+deb10u2~bpo9+1, liblttng-ust0-2.9.0-2+deb9u1, liblzma5-5.2.2-1.2+b1, libnettle6-3.3-1+b2, libnghttp2-14-1.18.1-1+deb9u1, libp11-kit0-0.23.3-2+deb9u1, libpcre3-2:8.39-3, libpsl5-0.17.0-3, librtmp1-2.4+20151223.gitfa8646d.1-1+b1, libsasl2-2-2.1.27~101-g0780600+dfsg-3+deb9u1, libselinux1-2.6-3+b3, libssh2-1-1.7.0-1+deb9u1, libssl1.0.2-1.0.2u-1~deb9u4, libtasn1-6-4.10-1.1+deb9u1, libunistring0-0.9.6+really0.9.3-0.1, libunwind8-1.1-4.1, libuuid1-2.29.2-1+deb9u1, zlib1g-1:1.2.8.dfsg-5
        Licenses found in Layer:  None
------------------------------------------------
```

</details>

File: tern/analyze/default/collect.py [hunk 3]
```diff
@@ -37,6 +37,18 @@ def get_snippet_list(invoke_step, work_dir=None, envs=None):
...
+        snippet_list = invoke_step.get('host')
+        # If environment variables exist, set them
+        if envs:
+            for var in envs:
+                snippet_list.insert(
+                    0, 'export ' + var.split('=')[0] + '=' + var.split('=')[1]
+                )
+        # If work_dir exist cd into it
+        if work_dir is not None:
+            snippet_list.insert(0, 'cd ' + work_dir)
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply tern/analyze/default/collect.py:50]: I can remove `env`, however work_dir is used to set the base path to the untar directory. As we're not chrooting into the layer, we still need to set the working directory.

**Comment by JamieMagee** [comment]
@nishakm I think I've addressed your comments. Ready for another round of review 👍 

I believe the failure in test setup is not related to my changes. It appears that a test is broken upstream.

**Comment by rnjudge** [comment]
> I believe the failure in test setup is not related to my changes. It appears that a test is broken upstream.

It is. I filed an [issue](https://github.com/spdx/tools-java/issues/14) for it and we can ignore it for now.

File: tern/analyze/default/core.py [hunk 4]
```diff
@@ -93,6 +95,46 @@ def execute_base(layer_obj, prereqs):
...
+    find packagesinstalled in the layer:
+        1. Use command_lib's base to to see if there is a method to
+            retrieve the metadata
+        2. If there is, invoke the scripts on the host and process
+            the results
+        3. Add the results to the ImageLayer object"""
+    # Add notices to this layer object
+    origin_layer = 'Layer {}'.format(layer_obj.layer_index)
+    # find the binary listing
+    listing = command_lib.get_base_listing("distroless")
```
Comment (nishakm) [inline tern/analyze/default/core.py:109]: I don't think you need this: `default_common.get_base_bin(layer)` will return `distroless` which will automatically go through the `execute_base` path.

**Review by nishakm — CHANGES_REQUESTED** [review]
One question about reimplementing `execute_base` for distroless. Everything else looks good!

File: tern/analyze/default/core.py (hunk 4 above)
Comment (nishakm) [reply tern/analyze/default/core.py:109]: Sorry: It looks like the reason there is a separate distroless execution path is because there is no shell.

**Comment by nishakm** [comment]
The linting fails because there are too many branches in `get_pkg_attrs`. I wonder if you could move the `invoke_in_rootfs` and `invoke_on_host` lines into a new function? That way the number of `if` statements in this function won't be that many and pylint will be happy.

File: tern/analyze/default/core.py (hunk 4 above)
Comment (JamieMagee) [reply tern/analyze/default/core.py:109]: Yeah, unfortunately distroless has to be a special case.

**Comment by JamieMagee** [comment]
I _think_ I have appeased pylint. I even remembered to sign-off my commit this time!

**Comment by nishakm** [comment]
@JamieMagee last request: can you squash your two commits and keep the first commit without the links to the external files? Thanks for continuing to work on this!

**Comment by JamieMagee** [comment]
@nishakm I've squashed the commits, but the commit message lint is failing due to the included URLs. I _think_ Linux Kernel commit messages have an exception to the 72 character limit for URLs, compiler messages, etc. But let me know what you want me to do. Thanks!

**Comment by nishakm** [comment]
@rnjudge Can we merge this one even though the link length is more than 72 characters? I think we can add the link exception to the commit message linter later.

**Comment by rnjudge** [comment]
> @rnjudge Can we merge this one even though the link length is more than 72 characters? I think we can add the link exception to the commit message linter later.

I am fine with this. Maybe @jaindhairyahere can add a check for this since he's working on something similar in https://github.com/tern-tools/tern/pull/911

**Review by nishakm — APPROVED** [review]

**Pull request merged**